python/requirements_test.txt

```diff
@@ -75,21 +76,46 @@ memray; platform_system != "Windows" and sys_platform != "darwin"
 # For doc tests
 myst-parser==0.15.2
 myst-nb==0.13.1
+sphinx==4.3.2
 jupytext==1.13.6
+jinja2==3.0.3
 pytest-docker-tools==3.1.3
 pytest-forked==1.4.0
 
 # For dataset tests
 polars==0.14.21
 
+# Dependencies from requirements.txt that are pinned
+# in library requirements
+pandas==1.3.5; python_version < '3.8'  # modin
+pandas==1.5.3; python_version >= '3.8'  # modin
+botocore==1.27.59
+protobuf==3.19.6
+typer==0.6.1
+uvicorn==0.16.0
+responses==0.13.4
+rich==12.0.1
+# ale-py requires 4.10.0. It's also compatible with flake8.
+# It's not compatible with mosaicml though
+importlib-metadata==4.10.0
```

Comment:
doesn't this result in discrepancy between what is installed in CI and what is in the docker images?

if these are pinned in `setup.py` then they should be pinned in `requirements.txt`. If not, then they should remain unpinned.

Reply:
We install a large number of dependencies in `requirements_test.py` that result in a mismatch between CI and Docker anyway. Without these pins, resolving backtracking dependencies times out (or errors out before).

The current PR is not about making both environments the same. That requires more work. This is explained in the PR description:

> Please note that this does not ensure we end up with the same python libraries in docker - in Docker, we don't install requirements_test.txt, so there are less constraints on the dependencies.

> However, this PR can be used to e.g. use pip-compile to compile a list of dependency constraints, and use these in the docker image build to achieve the same dependencies.

Many of these problems (and pins) will be gone when we upgrade CI to Python 3.8, and it will become much more feasible to make the environments similar / the same without any pins of core dependencies.